src/coreclr/jit/helperexpansion.cpp

```diff
@@ -443,7 +443,282 @@ PhaseStatus Compiler::fgExpandThreadLocalAccess()
+        targetSymbCondBB = fgNewBBFromTreeAfter(BBJ_COND, prevBb, gtCloneExpr(targetSymbolAddr), debugInfo);
+
+        GenTree* targetSymbolAddrVal = gtNewIndir(TYP_I_IMPL, targetSymbolAddr);
+        GenTree* targetSymbolNullCond =
+            gtNewOperNode(GT_EQ, TYP_INT, targetSymbolAddrVal, gtNewIconNode(0, TYP_I_IMPL));
+        targetSymbolNullCond = gtNewOperNode(GT_JTRUE, TYP_VOID, targetSymbolNullCond);
+        fgInsertStmtAfter(targetSymbCondBB, targetSymbCondBB->firstStmt(), fgNewStmtFromTree(targetSymbolNullCond));
+
+        GenTreeCall* classCtorRunHelperCall =
+            gtNewIndCallNode(gtNewIconHandleNode((size_t)classCtorRunHelper.addr, GTF_ICON_FTN_ADDR), TYP_I_IMPL);
+
+        // arg0: unused
+        classCtorRunHelperCall->gtArgs.PushBack(this, NewCallArg::Primitive(gtNewIconNode(0)));
+
+        // arg1: -1 (index of inlined storage)
+        classCtorRunHelperCall->gtArgs.PushBack(this, NewCallArg::Primitive(gtNewIconNode(-1)));
+
+        // arg2: NonGCStatics targetSymbol
+        classCtorRunHelperCall->gtArgs.PushBack(this, NewCallArg::Primitive(gtCloneExpr(targetSymbolAddr)));
+
+        fgMorphArgs(classCtorRunHelperCall);
+
+        GenTree* lazyCtorValueDef = gtNewStoreLclVarNode(finalLclNum, classCtorRunHelperCall);
+        lazyCtorBB = fgNewBBFromTreeAfter(BBJ_ALWAYS, targetSymbCondBB, lazyCtorValueDef, debugInfo, true);
+
+        fgRemoveRefPred(block, prevBb);
+        fgAddRefPred(targetSymbCondBB, prevBb);
+        fgAddRefPred(lazyCtorBB, targetSymbCondBB);
+        fgAddRefPred(block, lazyCtorBB);
+
+        lazyCtorBB->bbJumpDest = block;
+
+        // Inherit the weights
+        block->inheritWeight(prevBb);
+        targetSymbCondBB->inheritWeight(prevBb);
+
+        targetSymbCondBB->bbNatLoopNum = prevBb->bbNatLoopNum;
+        lazyCtorBB->bbNatLoopNum       = prevBb->bbNatLoopNum;
+
+        assert(BasicBlock::sameEHRegion(prevBb, targetSymbCondBB));
+        assert(BasicBlock::sameEHRegion(prevBb, lazyCtorBB));
+
+        // lazyCtorBB is the new previous block going forward
+        prevBb = lazyCtorBB;
+    }
+
+    // Block ops inserted by the split need to be morphed here since we are after morph.
+    // We cannot morph stmt yet as we may modify it further below, and the morphing
+    // could invalidate callUse.
+    while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+    {
+        fgMorphStmtBlockOps(block, newFirstStmt);
+        newFirstStmt = newFirstStmt->GetNextStmt();
+    }
+
+    if (TargetOS::IsWindows)
+    {
+#ifdef TARGET_64BIT
+        GenTree* tlsValue          = nullptr;
+        unsigned tlsLclNum         = lvaGrabTemp(true DEBUGARG("TLS access"));
+        lvaTable[tlsLclNum].lvType = TYP_I_IMPL;
+
+        // Mark this ICON as a TLS_HDL, codegen will use FS:[cns] or GS:[cns]
+        uint32_t offsetOfThreadLocalStoragePointer = 0;
+#ifdef _MSC_VER
+        offsetOfThreadLocalStoragePointer = offsetof(_TEB, ThreadLocalStoragePointer);
```

Comment:
This offset should not be hardcoded in the JIT like this. It should come through JIT/EE interface to allow cross-targeting, ideally as `CORINFO_THREAD_STATIC_BLOCKS_INFO::offsetOfThreadLocalStoragePointer`.

Reply:
Sure, but since the jitinterface is in C#, what is a good way to get this from C#? let me explore some options.

Reply:
`11 * PointerSize`

There are existing places where we do things like this, e.g. https://github.com/dotnet/runtime/blob/3ed1a7dbacb619462ec8b6859ae084467228177d/src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.RyuJit.cs#L30C1-L31

Reply:
hhm, wondering if its ok to hardcode `11` in jitinterface of C#?

Reply:
Yes, it is fine to hardcode this to support cross-compilation. (You would have the same problem with supporting cross-compilation in C/C++ - it would require some amount of hardcoding or copy&pasting as well.)